packages/govern-console/src/Providers/Permissions.tsx

```diff
@@ -41,40 +50,58 @@ type Role = {
   who: string
 }
 
-type PermissionsProviderProps = {
-  children: React.ReactNode
-}
-
-export default function PermissionsProvider({
-  children,
-}: PermissionsProviderProps) {
-  const [permissions, setPermissions] = useState({
+function permissionsFromRoles(
+  account: string,
+  roles: Role[],
+  permissions: Permissions = {
     execute: false,
     schedule: false,
     veto: false,
     challenge: false,
-  })
+  },
+): Permissions {
+  return roles.reduce(
+    (permissions: Permissions, role: Role) => {
+      const anyOrCurrentAccount =
+        addressesEqual(role.who, ETH_ANY_ADDRESS) ||
+        addressesEqual(role.who, account)
+
+      if (!anyOrCurrentAccount) {
+        return permissions
+      }
+
+      const roleName = KNOWN_QUEUE_ROLES.get(role.selector)
+      if (roleName && isPermissionsKey(roleName)) {
+        permissions[roleName] = true
+      }
+      return permissions
+    },
+    { ...permissions },
```

Comment:
A subtle change compared to previously is that we are now copying `permissions`, rather than modifying it. I think this is why [calling `populatePermissions()` in a render](https://github.com/aragon/govern/pull/261/files#diff-690b8b9e540016c91f87342dc92c7c89b5e33076aaf5c380eecc213b03cf16d1L26) could be done without infinite loop.

To prevent another infinite loop to happen, `permissions` is now handled by `useReducer()`.

Reply:
oh nice! Feels much cleaner.